core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```
@@ -0,0 +1,155 @@
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Renames index folders from {index.name} to {index.uuid}
+ */
+public class IndexFolderUpgrader {
+    private final NodeEnvironment nodeEnv;
+    private final Settings settings;
+    private final ESLogger logger = Loggers.getLogger(IndexFolderUpgrader.class);
+    private final MetaDataStateFormat<IndexMetaData> indexStateFormat = metaDataMetaDataStateFormat();
+
+    /**
+     * Creates a new upgrader instance
+     * @param settings node settings
+     * @param nodeEnv the node env to operate on
+     */
+    IndexFolderUpgrader(Settings settings, NodeEnvironment nodeEnv) {
+        this.settings = settings;
+        this.nodeEnv = nodeEnv;
+    }
+
+    /**
+     * Renames the index folder found in <code>source</code>
+     * if the renamed folder does not exist, otherwise no-op
+     */
+    void upgrade(final Index index, final Path source) throws IOException {
+        final Path target = source.resolveSibling(index.getUUID());
+        if (Files.exists(target)) {
+            // skip we upgraded already
+            // we assume no not-yet-upgraded file named {index_uuid} exists
+            logger.info("[{}] no upgrade needed - already upgraded", target);
+            return;
+        }
+        boolean success = false;
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+            success = true;
+        } catch (NoSuchFileException ignored) {
+            assert ignored.getFile().equals(source.toAbsolutePath().toString()) : "must be same as the source";
+            // thrown when the source is non-existent because the folder was renamed
+            // by another node (shared FS) after we checked if the target exists
+            assert Files.exists(target) : "must be upgraded by another node";
```

wow this is scary as shit I guess this means we are restarting multiple nodes at the same time in a full cluster restart. I think we can't do this neither support it on a shared FS. If we run into this situation we should fail and not swallow exceptions IMO